src/java/org/apache/cassandra/index/sai/disk/v3/CassandraDiskAnn.java

```diff
@@ -0,0 +1,202 @@
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import io.github.jbellis.jvector.disk.CachingGraphIndex;
+import io.github.jbellis.jvector.disk.CompressedVectors;
+import io.github.jbellis.jvector.disk.OnDiskGraphIndex;
+import io.github.jbellis.jvector.graph.GraphIndex;
+import io.github.jbellis.jvector.graph.GraphSearcher;
+import io.github.jbellis.jvector.graph.ListRandomAccessVectorValues;
+import io.github.jbellis.jvector.graph.NeighborSimilarity;
+import io.github.jbellis.jvector.graph.SearchResult.NodeScore;
+import io.github.jbellis.jvector.util.Bits;
+import io.github.jbellis.jvector.vector.VectorSimilarityFunction;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.QueryContext;
+import org.apache.cassandra.index.sai.disk.format.IndexComponent;
+import org.apache.cassandra.index.sai.disk.v1.PerIndexFiles;
+import org.apache.cassandra.index.sai.disk.v1.SegmentMetadata;
+import org.apache.cassandra.index.sai.disk.v1.postings.ReorderingPostingList;
+import org.apache.cassandra.index.sai.disk.vector.CassandraOnHeapGraph;
+import org.apache.cassandra.index.sai.disk.vector.JVectorLuceneOnDiskGraph;
+import org.apache.cassandra.index.sai.disk.vector.OnDiskOrdinalsMap;
+
+public class CassandraDiskAnn implements JVectorLuceneOnDiskGraph, AutoCloseable
+{
+    private static final Logger logger = Logger.getLogger(CassandraDiskAnn.class.getName());
+
+    private final OnDiskOrdinalsMap ordinalsMap;
+    private final CachingGraphIndex graph;
+    private final VectorSimilarityFunction similarityFunction;
+
+    // only one of these will be not null
+    private final CompressedVectors compressedVectors;
+    private final ListRandomAccessVectorValues uncompressedVectors;
+
+    public CassandraDiskAnn(SegmentMetadata.ComponentMetadataMap componentMetadatas, PerIndexFiles indexFiles, IndexContext context) throws IOException
+    {
+        similarityFunction = context.getIndexWriterConfig().getSimilarityFunction();
+
+        SegmentMetadata.ComponentMetadata termsMetadata = componentMetadatas.get(IndexComponent.TERMS_DATA);
+        graph = new CachingGraphIndex(new OnDiskGraphIndex<>(() -> indexFiles.termsData().createReader(), termsMetadata.offset));
+
+        long pqSegmentOffset = componentMetadatas.get(IndexComponent.PQ).offset;
+        try (var reader = indexFiles.pq().createReader())
+        {
+            reader.seek(pqSegmentOffset);
+            var containsCompressedVectors = reader.readBoolean();
+            if (containsCompressedVectors) {
+                compressedVectors = CompressedVectors.load(reader, reader.getFilePointer());
+                uncompressedVectors = null;
+            }
+            else
+            {
+                compressedVectors = null;
+                var vectors = cacheOriginalVectors(graph);
+                uncompressedVectors = new ListRandomAccessVectorValues(vectors, vectors.get(0).length);
```

Comment:
In practice it's capped at the max number of vectors before we write PQ (=1000), but you're right, it's redundant.

Since BFS=3 for M=16 is 4k, I don't think we need to bother special casing the graph caching either.